providers/azure_oidc.go

```diff
@@ -0,0 +1,194 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"github.com/pusher/oauth2_proxy/pkg/requests"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/pusher/oauth2_proxy/pkg/apis/sessions"
+	"github.com/pusher/oauth2_proxy/pkg/logger"
+	"golang.org/x/oauth2"
+)
+
+type AzureOIDCProvider struct {
+	*OIDCProvider
+	PermittedGroups []string
+	ExemptedEmails  []string
+}
+
+type AzureOIDCClaims struct {
+	*OIDCClaims
+	Groups            []string `json:"groups"`
+	PreferredUsername string   `json:"preferred_username"`
+}
+
+// NewAzureOIDCProvider initiates a new AzureOIDCProvider
+func NewAzureOIDCProvider(p *ProviderData) *AzureOIDCProvider {
+	provider := AzureOIDCProvider{
+		OIDCProvider: NewOIDCProvider(p),
+	}
+	provider.ProviderName = "Azure OpenID Connect"
+	if p.Scope == "" {
+		provider.Scope = "openid email profile"
+	}
+
+	if provider.ProfileURL == nil || provider.ProfileURL.String() == "" {
+		provider.ProfileURL = &url.URL{
+			Scheme: "https",
+			Host:   "graph.microsoft.com",
+			Path:   "/v1.0/me",
+		}
+	}
+
+	return &provider
+}
+
+func (p *AzureOIDCProvider) GetLoginURL(redirectURI, state string) string {
+	var a url.URL
+	a = *p.LoginURL
+	params, _ := url.ParseQuery(a.RawQuery)
```

Comment:
Do we want to ignore the error?

Reply:
Is there an option for an error if this is using empty url.URL variable (other than golang bug)?

Reply:
In this case, I would put the error into a variable and panic if it occurs. If I've understood correctly, the only way to get to an error case would be a programming error, so a panic seems reasonable for that IMO, WDYT @syscll?

Reply:
Sounds good to me